torch/fx/lazy_graph_module.py

```diff
@@ -0,0 +1,73 @@
+from contextlib import contextmanager
+
+from torch.fx import GraphModule
+from ._compatibility import compatibility
+
+_use_lazy_graph_module = False
+_force_skip_lazy_graph_module = False  # used in unit test to skip lazy graph module
+
+
+@compatibility(is_backward_compatible=False)
+@contextmanager
+def use_lazy_graph_module(should_use: bool):
+    try:
+        global _use_lazy_graph_module
+        prior = _use_lazy_graph_module
+        _use_lazy_graph_module = should_use and not _force_skip_lazy_graph_module
+        yield
+    finally:
+        _use_lazy_graph_module = prior
+
+
+@compatibility(is_backward_compatible=False)
+def get_graph_module_cls():
+    return LazyGraphModule if _use_lazy_graph_module else GraphModule
+
+
+@compatibility(is_backward_compatible=False)
+class LazyGraphModule(GraphModule):
+    @classmethod
+    def from_graphmodule(cls, gm: GraphModule):
+        if isinstance(gm, LazyGraphModule):
+            return gm
+        else:
+            return LazyGraphModule(gm, gm.graph)
+
+    def real_recompile(self):
+        if self._needs_recompile():
+            self._real_recompile()
+
+    @classmethod
+    def _needs_recompile(cls):
+        return cls.forward is cls._lazy_forward
+
+    def _lazy_forward(self, *args, **kwargs):
+        self._real_recompile()
+        assert not self._needs_recompile()
+
+        # call `__call__` rather than 'forward' since recompilation may
+        # install a wrapper for `__call__` to provide a customized error
+        # message.
+        return self(*args, **kwargs)
+
+    forward = _lazy_forward
+
+    def _real_recompile(self):
+        return super().recompile()
+
+    @classmethod
+    def recompile(cls):
+        cls.forward = cls._lazy_forward
```

Comment:
This is extremely suspicious. Calling recompile here modifies the CLASS, which will affect ALL FX graphs, not just the instance you are working with.

Reply:
It does not affect every Fx graph/GraphModule. The reason is because GraphModule has a special `__new__` method https://github.com/pytorch/pytorch/blob/bb28965924cd142a9d12f475fa646c45acb8639f/torch/fx/graph_module.py#L335  which creates a new subclass for every new Fx GraphModule instance.

Reply:
btw, GraphModule.recompile also update `cls.forward` directly.